### ISSUE TITLE
[x86, R2R] Jit/Regressions/JitBlue/DevDiv_278523 test failure

### ISSUE THREAD

**Issue by sandreenko** [dotnet/coreclr#11476]
This test contains such IL sequence:
```
// This sequence of IL repros the issue. Note that the `ldc.i8` is necessary (rather than an `ldc.i4` that is
// implicitly converted to a long byt the `add`).
ldloca.s V_1  <- native int
ldc.i8 4          <- int64
add                <- native int + int64
```

On x86 this sequence is not compiled and executed in CoreRun, because the method that contains it (Test64Bit) is called under condition that pointer size is 64 bits.

However, CrossGen has to compile it and it blows up on the assert that address should have native int size.

ECMA says that it is " an invalid CIL instruction" ([ecma](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-335.pdf#page328), page 328, Table III.2: Binary Numeric Operations).

Jit64 and RyuJit64 process it as correct sequence and execute this code without any problems.

I will add PEVerify and jit32 results soon.

**Comment by sandreenko** [comment]
PTAL @dotnet/jit-contrib .
What is a contract for " an invalid CIL instruction"? Should we throw an exception? Could we execute such code? Could we execute it differently on different architects?

**Comment by sandreenko** [comment]
PEVerify output:
```
[IL]: Error: [DevDiv_278523.exe : C::Test32Bit][offset 0x0000000C][found address of value 'T'] Expected numeric type on the stack.
[IL]: Error: [DevDiv_278523.exe : C::Test64Bit][offset 0x00000014][found address of value 'T'] Expected numeric type on the stack.
2 Error(s) Verifying DevDiv_278523.exe
```
So it shows the error for native int + int64.
Why does it complaint about native int + int32?

**Comment by sandreenko** [comment]
jit32 ngen fails with the same assert.

**Comment by CarolEidt** [comment]
> Why does it complaint about native int + int32?

It is a managed pointer (the address of a local), and although it is not invalid IL, it is unverifiable to add a managed pointer and an integer (so peverify appropriately gives an error).

> What is a contract for " an invalid CIL instruction"? Should we throw an exception?

We should invoke the `BADCODE` macro, which causes an exception to be thrown (by the VM I believe).

> Could we execute such code? Could we execute it differently on different architects?

In theory invalid IL should be independent of target architecture, but for historical reasons we have generally permitted IL like this that was previously permitted by the legacy 64-bit JIT.

**Comment by RussKeldorph** [comment]
@CarolEidt Would it be reasonable to make CoreCLR always invoke BADCODE for this and leave it as a (Full) .NET quirk?  It's too late IMO to introduce such a breaking change for x64 for 2.0, but perhaps we could see if it would float for a future release.  Seems like we should at least make sure the JIT64 behavior doesn't propagate to ARM64 and ARM right now.

**Comment by CarolEidt** [comment]
I believe that the following from Ecma 335 II.3, provides for the latitude to allow invalid IL to be either allowed or not:

> Validation refers to the application of a set of tests on any file to check that the file’s format, metadata, and CIL are self-consistent. These tests are intended to ensure that the file conforms to the normative requirements of this specification. When a conforming implementation of the CLI is presented with a non-conforming file, the behavior is unspecified.
...
> Aside from these rules, this standard leaves as unspecified:
> - The time at which (if ever) such an algorithm [verification] should be performed.
> - What a conforming implementation should do in the event of a verification failure.

Given the above, I believe that the right behavior is to invoke BADCODE for this on 32-bit targets, and continue to allow it on 64-bit systems.

**Comment by sandreenko** [comment]
Fixed in #13709.
